src/main/java/com/mongodb/kafka/connect/source/StartedMongoSourceTask.java

```diff
@@ -0,0 +1,648 @@
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaAndValue;
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.errors.DataException;
+import org.apache.kafka.connect.source.SourceRecord;
+import org.apache.kafka.connect.source.SourceTaskContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.bson.BsonDocument;
+import org.bson.BsonDocumentWrapper;
+import org.bson.Document;
+import org.bson.RawBsonDocument;
+
+import com.mongodb.MongoClientSettings;
+import com.mongodb.MongoCommandException;
+import com.mongodb.MongoException;
+import com.mongodb.client.ChangeStreamIterable;
+import com.mongodb.client.MongoChangeStreamCursor;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoDatabase;
+import com.mongodb.client.model.changestream.ChangeStreamDocument;
+import com.mongodb.lang.Nullable;
+
+import com.mongodb.kafka.connect.source.heartbeat.HeartbeatManager;
+import com.mongodb.kafka.connect.source.producer.SchemaAndValueProducer;
+import com.mongodb.kafka.connect.source.statistics.StatisticsManager;
+import com.mongodb.kafka.connect.source.topic.mapping.TopicMapper;
+import com.mongodb.kafka.connect.util.VisibleForTesting;
+import com.mongodb.kafka.connect.util.jmx.SourceTaskStatistics;
+import com.mongodb.kafka.connect.util.time.InnerOuterTimer;
+import com.mongodb.kafka.connect.util.time.InnerOuterTimer.InnerTimer;
+
+final class StartedMongoSourceTask {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MongoSourceTask.class);
+  private static final String FULL_DOCUMENT = "fullDocument";
+  private static final int NAMESPACE_NOT_FOUND_ERROR = 26;
+  private static final int ILLEGAL_OPERATION_ERROR = 20;
+  private static final int INVALIDATED_RESUME_TOKEN_ERROR = 260;
+  private static final int CHANGE_STREAM_FATAL_ERROR = 280;
+  private static final int CHANGE_STREAM_HISTORY_LOST = 286;
+  private static final int BSON_OBJECT_TOO_LARGE = 10334;
+  private static final Set<Integer> INVALID_CHANGE_STREAM_ERRORS =
+      new HashSet<>(
+          asList(
+              INVALIDATED_RESUME_TOKEN_ERROR,
+              CHANGE_STREAM_FATAL_ERROR,
+              CHANGE_STREAM_HISTORY_LOST,
+              BSON_OBJECT_TOO_LARGE));
+  private static final String RESUME_TOKEN = "resume token";
+  private static final String RESUME_POINT = "resume point";
+  private static final String NOT_FOUND = "not found";
+  private static final String DOES_NOT_EXIST = "does not exist";
+  private static final String INVALID_RESUME_TOKEN = "invalid resume token";
+  private static final String NO_LONGER_IN_THE_OPLOG = "no longer be in the oplog";
+
+  private final Supplier<SourceTaskContext> sourceTaskContextAccessor;
+  private final Time time;
+  private volatile boolean isRunning;
+  private boolean isCopying;
```

Comment:
`isCopying` is now a plain field because there is no point in setting it to `false` in `stop`. And without writes from `stop`, this field is always accessed sequentially.